dmscripts/export_dos_suppliers.py

```diff
@@ -0,0 +1,275 @@
+        return record['onFramework'] is True
+
+    def create_row(self, record):
+        return \
+            supplier_info(record) + \
+            contact_details(record) + \
+            service_counts(record)
+
+
+class FailedHandler(object):
+    NAME = 'failed'
+
+    def matches(self, record):
+        return record['onFramework'] is False
+
+    def create_row(self, record):
+        return \
+            supplier_info(record) + \
+            failed_mandatory_questions(record) + \
+            contact_details(record) + \
+            service_counts(record)
+
+
+class DiscretionaryHandler(object):
+    NAME = 'discretionary'
+
+    def matches(self, record):
+        return record['onFramework'] is None
+
+    def create_row(self, record):
+        return \
+            supplier_info(record) + \
+            discretionary_questions(record) + \
+            contact_details(record) + \
+            service_counts(record)
+
+
+class MultiCSVWriter(object):
+    """Manage writing to multiple CSV files"""
+    def __init__(self, output_dir, handlers):
+        self.output_dir = output_dir
+        self.handlers = handlers
+        self._csv_writers = dict()
+        self._csv_files = dict()
+        self._counters = {
+            handler.NAME: 0 for handler in handlers
+        }
+
+    def write_row(self, record):
+        for handler in self.handlers:
+            if handler.matches(record):
+                self._counters[handler.NAME] += 1
+                row = handler.create_row(record)
+                return self.csv_writer(handler, row).writerow(dict(row))
+        raise Exception("record not handled by any handler")
+
+    def csv_writer(self, handler, row):
+        if handler.NAME not in self._csv_writers:
+            fieldnames = [key for key, _ in row]
+            self._csv_writers[handler.NAME] = csv.DictWriter(self._csv_files[handler.NAME], fieldnames=fieldnames)
+            self._csv_writers[handler.NAME].writeheader()
+
+        return self._csv_writers[handler.NAME]
+
+    def csv_path(self, handler):
+        return os.path.join(self.output_dir, handler.NAME + '.csv')
+
+    def __enter__(self):
+        for handler in self.handlers:
+            self._csv_files[handler.NAME] = open(self.csv_path(handler), 'w+')
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        for f in self._csv_files.values():
+            f.close()
+
+    def print_counts(self):
+        print(" ".join("{}={}".format(handler.NAME, self._counters[handler.NAME]) for handler in self.handlers))
+
+
+def export_suppliers(client, content_loader, output_dir):
+    records = find_suppliers_with_details(client, content_loader, FRAMEWORK_SLUG)
+
+    handlers = [SuccessfulHandler(), FailedHandler(), DiscretionaryHandler()]
+
+    with MultiCSVWriter(output_dir, handlers) as writer:
+        for i, record in enumerate(records):
+            writer.write_row(record)
+            if i % 100 == 0:
```

Comment:
I like how we get an output line for the zeroth row :) - this confused me running against preview, but will be nice for when we run prod where there' s a lot more data.